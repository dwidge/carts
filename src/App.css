* {
box-sizing: border-box;
}

nav {
display: flex;
justify-content: space-between;
}

.link,.link-active {
color: inherit;
text-decoration: inherit;
border: solid 1px orange;
padding: 1em;
display: block;
}

.link-active {
background: orange;
}

button {
padding: 1em;
margin: 0em;
border: solid 1px orange;
}

textarea {
width: 100%;
height: 10em;
}

cart-table {
display: table;
width: 100%;
margin-top: 1em;
}
cart-header, cart-item {
display: table-row;
}
cart-header {
background: orange;
}
cart-header > *, cart-item > * {
display: table-cell;
vertical-align: middle;
width: 40%;
}
cart-header > *, cart-item > * {
padding: 0.5em;
}
cart-name > input {
width: 80%;
}
cart-buttons {
width: 20%;
padding: 0;
}
cart-buttons > button {
width: 100%;
}
