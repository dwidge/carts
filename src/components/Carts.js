import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { uuid, getItemById, replaceItemById, dropItemById, calcCsvFromObjects, calcObjectsFromCsv } from '@dwidge/lib'
import { onChange, onChangeChecks } from '@dwidge/lib-react'
import { ImportFile, ExportFile } from '@dwidge/table-react'

const Carts = ({ stslots, stcarts }) => {
	const [allslots] = stslots
	const [rawcarts, setcarts] = stcarts
	const [idEdit, setidEdit] = useState()
	const [confirm, setconfirm] = useState(false)

	const dropMissingSlots = cart => ({ ...cart, slots: cart.slots.filter(id => getItemById(allslots, id)) })
	const carts = rawcarts.map(dropMissingSlots)

	const setcart = cart => setcarts(replaceItemById(carts, cart))
	const delcart = id =>
		setcarts(dropItemById(carts, id))
	const addcart = cart => setcarts(carts.concat([cart]))
	const newcart = () => ({ id: uuid(), name: 'cart', slots: [] })

	const onClear = () => {
		confirm && setcarts([])
		setconfirm(!confirm)
	}

	const viewcarts = carts.map(({ id, name, slots }) => ({
		id,
		name,
		slots: slots.map(id => getItemById(allslots, id)).map(({ day, beg, end }) => `${day} ${beg}-${end}`),
		slotsId: slots,
	}))

	return (
		<div-page>
			<cart-table data-testid="cartTable">
				<cart-header>
					<cart-name>Name</cart-name>
					<cart-slots>Slots</cart-slots>
				</cart-header>
				{carts.map(cart => idEdit === cart.id ? (<CartEdit key={cart.id} allslots={allslots} cart={cart} onSave={cart => { setcart(cart); setidEdit() }} onCancel={setidEdit} />) : (<CartRow key={cart.id} allslots={allslots} cart={cart} onEdit={setidEdit} onDel={delcart} />))}
			</cart-table>
			<button onClick={() => addcart(newcart())} data-testid="buttonAdd">Add</button>
			<button onClick={onClear} data-testid="buttonClear">{confirm ? 'Confirm' : 'Clear'}</button>
			<ImportFile ext='.csv' onAccept={text => {
				const a = calcObjectsFromCsv(text).map(({ id, name, slotsId }) => ({
					id, name, slots: (slotsId.map) ? slotsId : [slotsId],
				}))
				setcarts(carts.concat(a))
			}}/>
			<ExportFile ext='.csv' name='carts.csv' content={calcCsvFromObjects(viewcarts)}/>
			<p>Name is any text or address to help you identify the cart.</p>
			<p>Slots are the allowed time slots for the cart which you pick from the Slots page.</p>
			<p>Examples</p>
			<p>Name: Cart A, Slots: 1 6-8, 7 14-16</p>
			<p>Name: Waterfall Rd, Slots: 3 6.75-8.25</p>
		</div-page>
	)
}

Carts.propTypes = {
	stslots: PropTypes.array.isRequired,
	stcarts: PropTypes.array.isRequired,
}

const CartRow = ({ allslots, cart, onEdit, onDel }) => {
	const { id, name, slots } = cart

	return (
		<cart-item>
			<cart-name>{name}</cart-name>
			<cart-slots>{slots.map(id => getItemById(allslots, id)).map(({ id, day, beg, end }) =>
				(<div key={id}>{day} {beg}-{end}</div>),
			)}</cart-slots>
			<cart-buttons>
				<button onClick={() => onEdit(id)} data-testid={'buttonEdit' + id}>Edit</button>
				<button onClick={() => onDel(id)} data-testid={'buttonDel' + id}>Del</button>
			</cart-buttons>
		</cart-item>
	)
}

CartRow.propTypes = {
	allslots: PropTypes.array.isRequired,
	cart: PropTypes.object.isRequired,
	onEdit: PropTypes.func.isRequired,
	onDel: PropTypes.func.isRequired,
}

const CartEdit = ({ allslots, cart, onSave, onCancel }) => {
	const { id } = cart
	const [name, setname] = useState(cart.name)
	const [slots, setslots] = useState(cart.slots)

	return (
		<cart-item>
			<cart-name><input data-testid="inputName" value={name} onChange={onChange(setname)} /></cart-name>
			<cart-slots>{allslots.map(({ id, day, beg, end }) =>
				(<div key={id}><input data-testid={'inputSlots' + id} type="checkbox" checked={slots.includes(id)} onChange={onChangeChecks(id, slots, setslots)} /> {day} {beg}-{end}</div>),
			)}</cart-slots>
			<cart-buttons>
				<button onClick={() => onSave({ id, name, slots })} data-testid="buttonSave">Save</button>
				<button onClick={onCancel} data-testid="buttonCancel">Cancel</button>
			</cart-buttons>
		</cart-item>
	)
}

CartEdit.propTypes = {
	allslots: PropTypes.array.isRequired,
	cart: PropTypes.object.isRequired,
	onSave: PropTypes.func.isRequired,
	onCancel: PropTypes.func.isRequired,
}

export default Carts
