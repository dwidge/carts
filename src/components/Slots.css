slot-table {
display: table;
width: 100%;
margin-top: 1em;
}
slot-header, slot-item {
display: table-row;
}
slot-header {
background: orange;
}
slot-header > *, slot-item > * {
display: table-cell;
vertical-align: middle;
}
slot-header > *, slot-item > * {
padding: 0.5em;
}
slot-item > * > input {
width: 80%;
}
slot-buttons {
width: 20%;
padding: 0;
}
slot-buttons > button {
width: 100%;
}
