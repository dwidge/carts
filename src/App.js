import React, { useState, useEffect } from 'react'
import {
	Routes,
	Route,
	NavLink,
} from 'react-router-dom'
import { Storage } from '@dwidge/lib-react'

import Slots from './components/Slots'
import Carts from './components/Carts'
import Pubs from './components/Pubs'
import Schedule from './components/Schedule'
import Data from './components/Data'
import './App.css'
const { useStorage } = Storage(useState, useEffect)

const App = () => {
	const stslots = useStorage('carts_slots', [])
	const stcarts = useStorage('carts_carts', [])
	const stpubs = useStorage('carts_pubs', [])

	return (
		<>
			<nav>
				<NavLink className={({ isActive }) => isActive ? 'link-active' : 'link'} to='/slots'>Slots</NavLink>
				<NavLink className={({ isActive }) => isActive ? 'link-active' : 'link'} to='/carts'>Carts</NavLink>
				<NavLink className={({ isActive }) => isActive ? 'link-active' : 'link'} to='/pubs'>Pubs</NavLink>
				<NavLink className={({ isActive }) => isActive ? 'link-active' : 'link'} to='/schedule'>Schedule</NavLink>
				<NavLink className={({ isActive }) => isActive ? 'link-active' : 'link'} to='/data'>Data</NavLink>
			</nav>
			<div>
				<Routes>
					<Route path='/' />
					<Route path='/slots' element={<Slots stslots={stslots} />}/>
					<Route path='/carts' element={<Carts stslots={stslots} stcarts={stcarts} />}/>
					<Route path='/pubs' element={<Pubs slots={stslots[0]} carts={stcarts[0]} stpubs={stpubs} />} />
					<Route path='/schedule' element={<Schedule slots={stslots[0]} carts={stcarts[0]} pubs={stpubs[0]} />}/>
					<Route path='/data' element={<Data {...{ stslots, stcarts, stpubs }} />}/>
				</Routes>
			</div>
		</>
	)
}

export default App
