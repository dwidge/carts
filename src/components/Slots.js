import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { uuid, replaceItemById, dropItemById, calcCsvFromObjects, calcObjectsFromCsv } from '@dwidge/lib'
import { onChange } from '@dwidge/lib-react'
import { ImportFile, ExportFile } from '@dwidge/table-react'

import './Slots.css'

const Slots = ({ stslots }) => {
	const [slots, setslots] = stslots
	const [idEdit, setidEdit] = useState()
	const [confirm, setconfirm] = useState(false)

	const setslot = slot => setslots(replaceItemById(slots, slot))
	const delslot = id =>
		setslots(dropItemById(slots, id))
	const addslot = slot => setslots(slots.concat([slot]))
	const newslot = () => ({ id: uuid(), day: 1, beg: 6, end: 8 })

	const onClear = () => {
		confirm && setslots([])
		setconfirm(!confirm)
	}

	return (
		<div-page>
			<slot-table data-testid="slotTable">
				<slot-header>
					<slot-day>Day</slot-day>
					<slot-beg>Beg</slot-beg>
					<slot-end>End</slot-end>
				</slot-header>
				{slots.map(slot => idEdit === slot.id ? (<SlotEdit key={slot.id} slot={slot} onSave={slot => { setslot(slot); setidEdit() }} onCancel={setidEdit} />) : (<SlotRow key={slot.id} slot={slot} onEdit={setidEdit} onDel={delslot} />))}
			</slot-table>
			<button onClick={() => addslot(newslot())} data-testid="buttonAdd">Add</button>
			<button onClick={onClear} data-testid="buttonClear">{confirm ? 'Confirm' : 'Clear'}</button>
			<ImportFile ext='.csv' onAccept={text => {
				const a = calcObjectsFromCsv(text).map(({ id, day, beg, end }) => ({
					id, day, beg, end,
				}))
				setslots(slots.concat(a))
			}}/>
			<ExportFile ext='.csv' name='slots.csv' content={calcCsvFromObjects(slots)}/>
			<p>Day is the number of the day in the week or beyond if the schedule covers multiple weeks. 1=Mon,7=Sun,8=2nd Mon</p>
			<p>Beg, End is the start and end hours in a 24 hour day. Use .5 for half past.</p>
			<p>Examples</p>
			<p>Day: 1, Beg: 6, End: 8</p>
			<p>Day: 7, Beg: 14, End: 16</p>
			<p>Day: 8, Beg: 12.5, End: 14.5</p>
		</div-page>
	)
}

Slots.propTypes = {
	stslots: PropTypes.array.isRequired,
}

const SlotRow = ({ slot, onEdit, onDel }) => {
	const { id, day, beg, end } = slot

	return (
		<slot-item>
			<slot-day>{day}</slot-day>
			<slot-beg>{beg}</slot-beg>
			<slot-end>{end}</slot-end>
			<slot-buttons>
				<button onClick={() => onEdit(id)} data-testid={'buttonEdit' + id}>Edit</button>
				<button onClick={() => onDel(id)} data-testid={'buttonDel' + id}>Del</button>
			</slot-buttons>
		</slot-item>
	)
}

SlotRow.propTypes = {
	slot: PropTypes.object.isRequired,
	onEdit: PropTypes.func.isRequired,
	onDel: PropTypes.func.isRequired,
}

const SlotEdit = ({ slot, onSave, onCancel }) => {
	const { id } = slot
	const [day, setday] = useState(slot.day)
	const [beg, setbeg] = useState(slot.beg)
	const [end, setend] = useState(slot.end)

	return (
		<slot-item>
			<slot-day><input data-testid="inputDay" value={day} onChange={onChange(setday)} /></slot-day>
			<slot-beg><input data-testid="inputBeg" value={beg} onChange={onChange(setbeg)} /></slot-beg>
			<slot-end><input data-testid="inputEnd" value={end} onChange={onChange(setend)} /></slot-end>
			<slot-buttons>
				<button onClick={() => onSave({ id, day, beg, end })} data-testid="buttonSave">Save</button>
				<button onClick={onCancel} data-testid="buttonCancel">Cancel</button>
			</slot-buttons>
		</slot-item>
	)
}
SlotEdit.propTypes = {
	slot: PropTypes.object.isRequired,
	onSave: PropTypes.func.isRequired,
	onCancel: PropTypes.func.isRequired,
}

export default Slots
