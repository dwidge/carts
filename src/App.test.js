import React, { useState } from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import * as J from '@dwidge/lib-react'
import { getTag, getTags, getText } from '@dwidge/lib-react'
import * as Lib from '@dwidge/lib'

import Slots from './components/Slots'
import Carts from './components/Carts'

import {
	MemoryRouter,
} from 'react-router-dom'
import App from './App'

const text = J.tools(userEvent, screen, jest).text
const type = J.type(userEvent, screen)
const input = J.input(userEvent, screen)
const click = J.click(userEvent, screen)
const serialSpy = J.serialSpy(jest)

jest.mock('@dwidge/lib', () => {
	return {
		__esModule: true,
		...jest.requireActual('@dwidge/lib'),
	}
})
beforeEach(async () => {
	serialSpy(Lib, 'uuid', [1, 2, 3])
})
afterEach(() => {
	jest.restoreAllMocks()
})

const getSlotList = () =>
	getTags(screen.getByTestId('slotTable'))('slot-item').map(slotItem => ['slot-day', 'slot-beg', 'slot-end'].map(getTag(slotItem)).map(getText))

const getCartList = () =>
	getTags(screen.getByTestId('cartTable'))('cart-item').map(item => [getText(getTag(item)('cart-name')), getTags(getTag(item)('cart-slots'))('div').map(getText)])

describe('Slots', () => {
	const slot1 = { id: 1, day: 1, beg: 6, end: 8 }
	const slot2 = { id: 2, day: 2, beg: 14, end: 17 }

	it('enters day,beg,end into list', async () => {
		const App = () => (<Slots stslots={useState([])}/>)
		render(<App/>)
		click('buttonAdd')
		click('buttonEdit1')
		await input('inputDay', '1')
		await input('inputBeg', '6')
		await input('inputEnd', '8')
		click('buttonSave')
		expect(getSlotList()).toEqual([['1', '6', '8']])
	})

	it('enters another day,beg,end into list', async () => {
		const App = () => (<Slots stslots={useState([slot1])}/>)
		render(<App/>)
		expect(getSlotList()).toEqual([['1', '6', '8']])
		Lib.uuid()
		click('buttonAdd')
		click('buttonEdit2')
		await input('inputDay', '2')
		await input('inputBeg', '14')
		await input('inputEnd', '17')
		click('buttonSave')
		expect(getSlotList()).toEqual([['1', '6', '8'], ['2', '14', '17']])
	})

	it('confirms & clears list', async () => {
		const App = () => (<Slots stslots={useState([slot1, slot2])}/>)
		render(<App/>)
		expect(getSlotList()).toEqual([['1', '6', '8'], ['2', '14', '17']])
		expect(text('buttonClear')).toEqual('Clear')
		click('buttonClear')
		expect(text('buttonClear')).toEqual('Confirm')
		expect(getSlotList()).toEqual([['1', '6', '8'], ['2', '14', '17']])
		click('buttonClear')
		expect(text('buttonClear')).toEqual('Clear')
		expect(getSlotList()).toEqual([])
	})
})

describe('Carts', () => {
	const slot1 = { id: 1, day: 1, beg: 6, end: 8 }
	const cart1 = { id: 1, name: 'cart1', slots: [1] }
	const cart2 = { id: 2, name: 'cart2', slots: [] }

	it('enters cart into list', async () => {
		const App = () => (<Carts stslots={useState([])} stcarts={useState([])} />)
		render(<App/>)
		click('buttonAdd')
		click('buttonEdit1')
		await type('inputName', '1')
		click('buttonSave')
		expect(getCartList()).toEqual([['cart1', []]])
	})

	it('enables a slot', async () => {
		const App = () => (<Carts stslots={useState([slot1])} stcarts={useState([])} />)
		render(<App/>)
		click('buttonAdd')
		click('buttonEdit1')
		await type('inputName', '1')
		click('inputSlots1')
		click('buttonSave')
		expect(getCartList()).toEqual([['cart1', ['1 6-8']]])
	})

	it('clears missing slot', async () => {
		const App = () => (<Carts stslots={useState([])} stcarts={useState([cart1])} />)
		render(<App/>)
		expect(getCartList()).toEqual([['cart1', []]])
	})

	it('confirms & clears list', async () => {
		const App = () => (<Carts stslots={useState([slot1])} stcarts={useState([cart1, cart2])} />)
		render(<App/>)
		expect(getCartList()).toEqual([['cart1', ['1 6-8']], ['cart2', []]])
		expect(text('buttonClear')).toEqual('Clear')
		click('buttonClear')
		expect(text('buttonClear')).toEqual('Confirm')
		expect(getCartList()).toEqual([['cart1', ['1 6-8']], ['cart2', []]])
		click('buttonClear')
		expect(text('buttonClear')).toEqual('Clear')
		expect(getCartList()).toEqual([])
	})
})

test('renders buttons', () => {
	render(<MemoryRouter initialEntries={['/carts']}>
		<App />
	</MemoryRouter>)

	expect(screen.getByText(/Add/i)).toBeInTheDocument()
	expect(screen.getByText(/Clear/i)).toBeInTheDocument()
})
describe('pubs', () => {
	test('renders buttons', () => {
		render(<MemoryRouter initialEntries={['/pubs']}>
			<App />
		</MemoryRouter>)

		expect(screen.getByText(/Add/i)).toBeInTheDocument()
		expect(screen.getByText(/Clear/i)).toBeInTheDocument()
	})
})
