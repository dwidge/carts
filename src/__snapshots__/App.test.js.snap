// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Data renders page 1`] = `
<div-page
  data-testid="pageData"
>
  <p>
    Slots: 
    0
  </p>
  <p>
    Carts: 
    0
  </p>
  <p>
    Pubs: 
    0
  </p>
  <button>
    Clear All
  </button>
  <div>
    <div>
      <button
        data-testid="buttonImport.json"
      >
        Import 
        .json
      </button>
    </div>
    <button>
      Export .json
    </button>
    <button>
      Export .csv
    </button>
  </div>
  <p>
    Import/export .json to backup or transfer between devices. Export .csv to use with a spreadsheet program.
  </p>
</div-page>
`;

exports[`App Schedule renders columns 1`] = `
<cart-table
  data-testid="tableSchedule"
>
  <cart-header>
    <cart-name>
      Cart
    </cart-name>
    <cart-slots>
      Slot
    </cart-slots>
    <cart-pub>
      Pub
    </cart-pub>
  </cart-header>
</cart-table>
`;

exports[`App Schedule renders columns 2`] = `
<cart-table
  data-testid="tableStats"
>
  <cart-header>
    <cart-name>
      Pub
    </cart-name>
    <cart-name>
      Slots
    </cart-name>
    <cart-name>
      Hours
    </cart-name>
  </cart-header>
</cart-table>
`;

exports[`Schedule shows schedule 1`] = `
<cart-table
  data-testid="tableSchedule"
>
  <cart-header>
    <cart-name>
      Cart
    </cart-name>
    <cart-slots>
      Slot
    </cart-slots>
    <cart-pub>
      Pub
    </cart-pub>
  </cart-header>
  <cart-item>
    <cart-name>
      cart1
    </cart-name>
    <cart-slots>
      1 6-8
    </cart-slots>
    <cart-pub>
      pub1
    </cart-pub>
  </cart-item>
</cart-table>
`;

exports[`Schedule shows schedule 2`] = `
<cart-table
  data-testid="tableStats"
>
  <cart-header>
    <cart-name>
      Pub
    </cart-name>
    <cart-name>
      Slots
    </cart-name>
    <cart-name>
      Hours
    </cart-name>
  </cart-header>
  <cart-item>
    <cart-name>
      pub1
    </cart-name>
    <cart-name>
      1
    </cart-name>
    <cart-name>
      2
    </cart-name>
  </cart-item>
  <cart-item>
    <cart-name>
      pub2
    </cart-name>
    <cart-name>
      0
    </cart-name>
    <cart-name>
      0
    </cart-name>
  </cart-item>
</cart-table>
`;
